src/preact/component/3p-frame.js

```diff
@@ -0,0 +1,144 @@
+import {generateSentinel, getBootstrapUrl} from '../../3p-frame';
+import {
+  getOptionalSandboxFlags,
+  getRequiredSandboxFlags,
+} from '../../core/3p-frame';
+import {parseUrlDeprecated} from '../../url';
+import {sequentialIdGenerator} from '../../utils/id-generator';
+import {useCallback, useEffect, useRef, useState} from '..';
+
+/** @type {!Object<string,function>} 3p frames for that type. */
+export const countGenerators = {};
+
+/** @enum {string} */
+const MessageType = {
+  // TODO(wg-bento): Add more types as they become needed.
+  EMBED_SIZE: 'embed-size',
+};
+
+// Block synchronous XHR in ad. These are very rare, but super bad for UX
+// as they block the UI thread for the arbitrary amount of time until the
+// request completes.
+const BLOCK_SYNC_XHR = "sync-xhr 'none';";
+
+// TODO(wg-bento): UA check for required flags without iframe element
+const DEFAULT_SANDBOX =
+  getRequiredSandboxFlags().join(' ') +
+  ' ' +
+  getOptionalSandboxFlags().join(' ');
+
+/**
+ * Creates the iframe for the embed. Applies correct size and passes the embed
+ * attributes to the frame via JSON inside the fragment.
+ * @param {!IframeDef.Props} props
+ * @return {PreactDef.Renderable}
+ */
+export function ProxyIframeEmbed({
+  name: nameProp,
+  options = {},
+  sandbox = DEFAULT_SANDBOX,
+  src,
+  type,
+  title = type,
+  ...rest
+}) {
+  const ref = useRef(null);
+  const contentRef = useRef(null);
+  if (!countGenerators[type]) {
+    countGenerators[type] = sequentialIdGenerator();
+  }
+  const [count] = useState(countGenerators[type]);
+  const [name, setName] = useState(nameProp);
+  useEffect(() => {
+    if (nameProp) {
+      setName(nameProp);
+      return;
+    }
+    const win = contentRef.current?.ownerDocument?.defaultView;
+    if (!win) {
+      return;
+    }
+    const attrs = dict({
+      'title': title,
+      'type': type,
+      '_context': dict({
+        'location': {
+          'href': win.location.href,
+        },
+        'sentinel': generateSentinel(win),
+      }),
+    });
+    for (const key in options) {
+      attrs[key] = options[key];
+    }
+    setName(
+      JSON.stringify(
+        dict({
+          'host': parseUrlDeprecated(src).hostname,
+          'bootstrap': getBootstrapUrl(),
+          'type': type,
+          // https://github.com/ampproject/amphtml/pull/2955
+          'count': count,
+          'attributes': attrs,
+        })
+      )
+    );
+  }, [count, nameProp, options, src, title, type]);
+
+  const manageMessageHandler = useCallback((ref, onSuccess) => {
+    const iframe = ref.current;
+    if (!iframe) {
+      return;
+    }
+    const messageHandler = (event) => {
+      const iframe = ref.current;
+      if (!iframe || event.source != iframe.contentWindow || !event.data) {
+        return;
+      }
+
+      const data = deserializeMessage(event.data);
+      if (data['type'] == MessageType.EMBED_SIZE) {
+        const height = data['height'];
+        onSuccess(height);
```

Comment:
Same as the comment on `IframeEmbed`. I think each iframe, for now, should process its own resize events.